plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java

```diff
@@ -339,10 +369,25 @@ public Map<String, BlobContainer> children() throws IOException {
     }
 
     private static List<ListObjectsV2Response> executeListing(AmazonS3Reference clientReference, ListObjectsV2Request listObjectsRequest) {
+        return executeListing(clientReference, listObjectsRequest, -1);
+    }
+
+    private static List<ListObjectsV2Response> executeListing(
+        AmazonS3Reference clientReference,
+        ListObjectsV2Request listObjectsRequest,
+        int limit
+    ) {
         return SocketAccess.doPrivileged(() -> {
             final List<ListObjectsV2Response> results = new ArrayList<>();
+            int totalObjects = 0;
             ListObjectsV2Iterable listObjectsIterable = clientReference.get().listObjectsV2Paginator(listObjectsRequest);
-            listObjectsIterable.forEach(results::add);
+            for (ListObjectsV2Response listObjectsV2Response : listObjectsIterable) {
+                results.add(listObjectsV2Response);
+                totalObjects += listObjectsV2Response.contents().size();
+                if (limit != -1 && totalObjects > limit) {
+                    break;
+                }
+            }
```

Comment:
Should we return the desired limit here itself else this method looks inconsistent(although not public) in the sense it can return more than the desired limit

Reply:
This won't be possible in this method. `ListObjectsV2Response` contains a list of keys which represents one page.